birdhouse/components/thredds/default.env

```diff
@@ -17,7 +17,25 @@ export THREDDS_SERVICE_DATA_LOCATION_NAME='Birdhouse'
 export THREDDS_DATASET_URL_PATH='datasets'
 export THREDDS_SERVICE_DATA_URL_PATH='birdhouse'
 
+export THREDDS_MAGPIE_EXTRA_METADATA_PREFIXES='".+\\.txt", ".+\\.md", ".+\\.rst"'
+export THREDDS_MAGPIE_EXTRA_DATA_PREFIXES=''
 
+export THREDDS_SERVICE_DATA_EXTRA_FILE_FILTERS=''
```

Comment:
How about using explicitly: 
```env
export THREDDS_MAGPIE_METADATA_PREFIXES='
[
  null,              # note: special YAML value evaluated as `no-prefix`, use quotes if literal value is needed
  "\\w+\\.gif",      # threddsIcon, folder icon, etc.
  "\\w+\\.ico",      # favicon
  "\\w+\\.css",      # tds.css
  "catalog\\.\\w+",  # note: special case for `THREDDS` top-level directory (root) accessed for `BROWSE`
  catalog,
  ncml,
  uddc,
  iso,
]
'
export THREDDS_MAGPIE_DATA_PREFIXES='
[
  fileServer,
  dodsC,
  wcs,
  wms,
  ncss,
]
'
```

With those variables defined by default next to the below `THREDDS_DATASET_DATASETSCAN_BODY`, it is easy for a user that desires to extend the capabilities to override the variables by a simple copy-paste. The doc can also more easily describe that overriding `THREDDS_MAGPIE_DATA_PREFIXES` for example would require the corresponding filter to be added to `THREDDS_DATASET_DATASETSCAN_BODY`. 

The issue I have with the current definitions is that modifying the `THREDDS_DATASET_DATASETSCAN_BODY` variable to remove the defaults will not reflect in the Magpie config unless it is completely overridden.

Reply:
But we don't want users to remove any of the existing metadata prefixes or it will break THREDDS.

Metadata prefixes:
- `null`: without this they can't access `example.com/thredds/`
- `.gif`, `.ico`, `.css`: without this, the catalog pages won't render properly
- `catalog\\.\\w+`, `catalog`: without this we can't view any of the catalog.html or catalog.xml pages (see catalog services https://docs.unidata.ucar.edu/tds/current/userguide/services_ref.html#catalog-services)
- the rest: these are metadata services offered by thredds https://docs.unidata.ucar.edu/tds/current/userguide/services_ref.html#nciso-metadata-services

Data prefixes:
- these are data services offered by thredds https://docs.unidata.ucar.edu/tds/current/userguide/services_ref.html#data-access-services

If the user changes any of those they either can't access crucial parts of the THREDDS catalog or they access metadata as data or vice versa. I don't think we want to let users do that.

> The doc can also more easily describe that overriding THREDDS_MAGPIE_DATA_PREFIXES for example would require the corresponding filter to be added to THREDDS_DATASET_DATASETSCAN_BODY

But if they modify filters in `THREDDS_DATASET_DATASETSCAN_BODY` they don't have to update `THREDDS_MAGPIE_DATA_PREFIXES` since new file types can still be accessed by the same THREDDS services that are described by the existing prefixes. They only need to modify `THREDDS_MAGPIE_DATA_PREFIXES` if:

- they create a brand new THREDDS service with a different prefix (which we discourage)
- they want a specific file extension to always be treated as metadata or as data regardless of the THREDDS service they are using to access the data.

In both of these cases, they only will ever need to add more fields, not remove or modify existing ones.

> The issue I have with the current definitions is that modifying the THREDDS_DATASET_DATASETSCAN_BODY variable to remove the defaults will not reflect in the Magpie config unless it is completely overridden.

But it does not need to be overridden. 

For example, if we remove the ncml service from `THREDDS_DATASET_DATASETSCAN_BODY` then a user trying to access a file using this service will no longer see the associated ncml URL as an option in the catalog. This does not affect the magpie permissions since the URL that would trigger the relevant Magpie permissions is no longer being offered by THREDDS.

Another example, if we change the filters in `THREDDS_DATASET_DATASETSCAN_BODY` to include files with a `.json` extension then we also don't need to modify anything in Magpie because the current magpie configuration already covers the case where we want to view the file in the catalog (the metadata `catalog` prefix) and where we want to access the file as data (the data `fileServer` prefix).

As usual, please let me know if I've misunderstood something about how Magpie works.

Reply:
Good point about the `catalog` and above items for rendering THREDDS browsing. Those can be left. 

For the others, they are relevant as hardcoded services because they happen to be enabled in https://github.com/bird-house/birdhouse-deploy/blob/master/birdhouse/components/thredds/threddsConfig.xml.template

So, I was wondering, if anyone is trying to modify which services/files are available or not by the various variables, isn't there a chance they are also overriding that config to modify which file extensions are accessible by certain services or not, to avoid weird side-effects?

For example, if enabling `.h5`, is it really relevant for this to be accessed via `wms`, `wcs`, etc.? Anything else than `fileServer` will most probably cause random errors.

_**Note**_: I do not have a strong feeling about this proposal. Modifying enabled services is a strong configuration change that I'm not planing to do. I was just shooting out the idea that if someone does try to modify them, they probably need to override almost everything anyway, and it's probably easier to do so with full-overrides rather than many concatenations of disparate variables. That is maybe out of scope for this PR.

Reply:
> For example, if enabling .h5, is it really relevant for this to be accessed via wms, wcs, etc.? Anything else than fileServer will most probably cause random errors.

Good question. In this example they would likely create a configuration that looks something like this:

```xml
<metadata inherited="true">
  <serviceName>fileServer</serviceName>
</metadata>

<filter>
  <include wildcard="*.h5" />
</filter>
```

This tells the `datasetScan` block to only serve files with an .h5 extension and only serve those with the `fileServer` service. 

This interacts with Magpie in the following ways:

- if a user tries to view the catalog for these files it will follow the Magpie rules for the `catalog\\.\\w+` prefix (browse permissions)
- if a user tries to access an .h5 file it will follow the Magpie rules for viewing the `fileServer` prefix (read permissions)
- if the user tries to access an .h5 file with another service, let's say `wms`, Magpie will follow the rules for viewing the `wms` prefix but then THREDDS will return a 404 error since there is no valid `wms` URL for the file.

The only time we need to modify the Magpie configuration is if we change the services themselves (which we discourage). However, with this change, that is technically possible by updating the  `THREDDS_MAGPIE_EXTRA_METADATA_PREFIXES` and `THREDDS_MAGPIE_EXTRA_DATA_PREFIXES` variables.

To summarize ... I don't think that users will really have to modify the Magpie configuration very much even if they're modifying the THREDDS configuration.

Reply:
Ok. Sounds good. Let's keep it the way it is currently.